test/contracts/wormholeJoin.ts

```diff
@@ -0,0 +1,137 @@
+import { MainnetSdk } from '@dethcrypto/eth-sdk-client'
+import { BigNumber, BigNumberish, Signer } from 'ethers'
+import { hexlify, hexZeroPad } from 'ethers/lib/utils'
+import { ethers } from 'hardhat'
+import { Dictionary } from 'ts-essentials'
+
+import {
+  L1AddWormholeDomainSpell__factory,
+  L1ConfigureWormholeSpell__factory,
+  WormholeConstantFee,
+  WormholeConstantFee__factory,
+  WormholeJoin,
+  WormholeJoin__factory,
+  WormholeOracleAuth,
+  WormholeOracleAuth__factory,
+  WormholeRouter,
+  WormholeRouter__factory,
+} from '../../typechain'
+import { getContractFactory, impersonateAccount, waitForTx } from '../helpers'
+import { BaseBridgeSdk } from './bridge'
+import { executeSpell } from './spell'
+
+const bytes32 = ethers.utils.formatBytes32String
+
+export const OPTIMISTIC_ROLLUP_FLUSH_FINALIZATION_TIME = 60 * 60 * 24 * 8 // flush should happen more or less, 1 day after initWormhole, and should take 7 days to finalize
+
+export async function deployWormhole({
+  defaultSigner,
+  sdk,
+  ilk,
+  joinDomain,
+  globalFee,
+  globalFeeTTL,
+}: {
+  defaultSigner: Signer
+  sdk: MainnetSdk
+  ilk: string
+  joinDomain: string
+  globalFee: BigNumberish
+  globalFeeTTL: BigNumberish
+}): Promise<{
+  join: WormholeJoin
+  oracleAuth: WormholeOracleAuth
+  router: WormholeRouter
+  constantFee: WormholeConstantFee
+}> {
+  const WormholeJoinFactory = getContractFactory<WormholeJoin__factory>('WormholeJoin', defaultSigner)
+  const join = await WormholeJoinFactory.deploy(sdk.vat.address, sdk.dai_join.address, ilk, joinDomain)
+  console.log('WormholeJoin deployed at: ', join.address)
+
+  const ConstantFeeFactory = getContractFactory<WormholeConstantFee__factory>('WormholeConstantFee', defaultSigner)
+  const constantFee = await ConstantFeeFactory.deploy(globalFee, globalFeeTTL)
+  console.log('ConstantFee deployed at: ', constantFee.address)
+
+  const WormholeOracleAuthFactory = getContractFactory<WormholeOracleAuth__factory>('WormholeOracleAuth', defaultSigner)
+  const oracleAuth = await WormholeOracleAuthFactory.deploy(join.address)
+
+  console.log('Deploying router...')
+  const WormholeRouterFactory = getContractFactory<WormholeRouter__factory>('WormholeRouter', defaultSigner)
+  const router = await WormholeRouterFactory.deploy(sdk.dai.address)
+
+  console.log('Finalizing permissions...')
```

Comment:
Why aren't the following lines included in `L1ConfigureWormholeSpell.sol` ?

Reply:
So my line of thinking was that outside of the spell we should:
* deploy
* deny deployer (this prevents deployer from doing something funny with contracts before spell is executed)